[2024-09-12 20:41:44,020] [INFO] [real_accelerator.py:158:get_accelerator] Setting ds_accelerator to cuda (auto detect)
/workspace/Oobleck/oobleck/__init__.py
node_index: 0
[2024-09-12 20:41:47,218] [WARNING] [agent.py:154:_launch_workers] Profile data for model gpt2 not found. Exception Error parsing json file: /workspace/Oobleck/tmp/profiles/gpt3_1_3B-16-13-1/mb16.json Launching profiler...
[2024-09-12 20:41:49,130] [INFO] [real_accelerator.py:158:get_accelerator] Setting ds_accelerator to cuda (auto detect)
[2024-09-12 20:41:52,401] [INFO] [profiler.py:270:profile] Profiling model gpt2. rank 0, world_size 13, num_workers_per_node 1.
Dataset: model_name: gpt2, dataset_path: wikitext, dataset_name: wikitext-2-raw-v1, max_seq_length: 1024
[2024-09-12 20:41:54,267] [INFO] [profiler.py:276:profile] Dataset loaded
model_name: gpt2, model_tag: gpt3_1_3B
[2024-09-12 20:41:55,309] [INFO] [profiler.py:284:profile] model loaded
[2024-09-12 20:42:18,559] [INFO] [profiler.py:292:profile] init pg. master_addr: 172.21.0.42, master_port: 23456
[2024-09-12 20:42:21,918] [INFO] [profiler.py:306:profile] Profiling model execution latency.
[2024-09-12 20:42:21,918] [INFO] [profiler.py:54:profile_execution_layers] Profiling layer execution latency: 0 iteration
[2024-09-12 20:42:27,089] [INFO] [profiler.py:54:profile_execution_layers] Profiling layer execution latency: 1 iteration
[2024-09-12 20:42:31,528] [INFO] [profiler.py:112:profile_execution_layers] before barrier
[2024-09-12 20:42:32,165] [INFO] [profiler.py:114:profile_execution_layers] after barrier
[2024-09-12 20:42:32,165] [INFO] [profiler.py:120:profile_execution_layers] before broadcast
[2024-09-12 20:42:32,165] [INFO] [profiler.py:122:profile_execution_layers] after broadcast
[2024-09-12 20:42:32,166] [INFO] [profiler.py:317:profile] Profiling cross-node allreduce latency.
[2024-09-12 20:42:32,166] [INFO] [profiler.py:160:profile_allreduce_across_nodes] Profile allreduce across 13 nodes latency
[2024-09-12 20:42:32,441] [INFO] [profiler.py:326:profile] Profiling in-node allreduce latency.
[2024-09-12 20:42:32,441] [INFO] [profiler.py:209:profile_allreduce_in_node] Profile allreduce within a node latency
[2024-09-12 20:42:33,462] [INFO] [agent.py:189:_launch_workers] in agent. my_ip_port 172.21.0.42:2220, node_ip_ports ['172.21.0.42:2220', '172.21.0.42:2221', '172.21.0.42:2222', '172.21.0.42:2223', '172.21.0.46:2220', '172.21.0.46:2221', '172.21.0.46:2222', '172.21.0.46:2223', '172.21.0.47:2220', '172.21.0.47:2221', '172.21.0.47:2222', '172.21.0.47:2223', '172.21.0.90:2220']
[2024-09-12 20:42:33,462] [INFO] [agent.py:192:_launch_workers] Launching worker 0...
[2024-09-12 20:42:35,399] [INFO] [real_accelerator.py:158:get_accelerator] Setting ds_accelerator to cuda (auto detect)
[2024-09-12 20:42:38,717] [INFO] [worker.py:23:worker_main] Initializing Oobleck Engine...
[2024-09-12 20:42:38,717] [INFO] [worker.py:24:worker_main] in worker main: my_ip 172.21.0.42:2220
/root/miniconda3/envs/oobleck/lib/python3.10/site-packages/transformers/training_args.py:1281: FutureWarning: using `no_cuda` is deprecated and will be removed in version 5.0 of ðŸ¤— Transformers. Use `use_cpu` instead
  warnings.warn(
Dataset: model_name: gpt2, dataset_path: wikitext, dataset_name: wikitext-2-raw-v1, max_seq_length: 1024
[2024-09-12 20:42:40,620] [INFO] [engine.py:635:_initialize_engine] model arguments: {'n_embd': 2048, 'n_head': 32, 'n_positions': 1024, 'num_hidden_layers': 24}
model_name: gpt2, model_tag: gpt3_1_3B
mb size: 26, ar across: 26, ar in: 26
Loading done. creating layer execution results...
Returning from get_profiler_results
[2024-09-12 20:42:41,607] [INFO] [engine.py:677:_initialize_engine] Number of nodes range: (2, 13)
min_nodes: 2, max_nodes: 13, gpus: 1, 
Create pipeline templates
StageExecutionResult[0:13] with 1 devices
StageExecutionResult[14:25] with 1 devices
StageExecutionResult[0:9] with 1 devices
StageExecutionResult[10:17] with 1 devices
StageExecutionResult[18:25] with 1 devices
StageExecutionResult[0:7] with 1 devices
StageExecutionResult[8:14] with 1 devices
StageExecutionResult[15:20] with 1 devices
StageExecutionResult[21:25] with 1 devices
StageExecutionResult[0:6] with 1 devices
StageExecutionResult[7:11] with 1 devices
StageExecutionResult[12:16] with 1 devices
StageExecutionResult[17:21] with 1 devices
StageExecutionResult[22:25] with 1 devices
StageExecutionResult[0:5] with 1 devices
StageExecutionResult[6:10] with 1 devices
StageExecutionResult[11:14] with 1 devices
StageExecutionResult[15:18] with 1 devices
StageExecutionResult[19:22] with 1 devices
StageExecutionResult[23:25] with 1 devices
StageExecutionResult[0:4] with 1 devices
StageExecutionResult[5:8] with 1 devices
StageExecutionResult[9:12] with 1 devices
StageExecutionResult[13:16] with 1 devices
StageExecutionResult[17:20] with 1 devices
StageExecutionResult[21:23] with 1 devices
StageExecutionResult[24:25] with 1 devices
StageExecutionResult[0:4] with 1 devices
StageExecutionResult[5:8] with 1 devices
StageExecutionResult[9:11] with 1 devices
StageExecutionResult[12:14] with 1 devices
StageExecutionResult[15:17] with 1 devices
StageExecutionResult[18:20] with 1 devices
StageExecutionResult[21:23] with 1 devices
StageExecutionResult[24:25] with 1 devices
StageExecutionResult[0:3] with 1 devices
StageExecutionResult[4:6] with 1 devices
StageExecutionResult[7:9] with 1 devices
StageExecutionResult[10:12] with 1 devices
StageExecutionResult[13:15] with 1 devices
StageExecutionResult[16:18] with 1 devices
StageExecutionResult[19:21] with 1 devices
StageExecutionResult[22:23] with 1 devices
StageExecutionResult[24:25] with 1 devices
StageExecutionResult[0:3] with 1 devices
StageExecutionResult[4:6] with 1 devices
StageExecutionResult[7:9] with 1 devices
StageExecutionResult[10:12] with 1 devices
StageExecutionResult[13:14] with 1 devices
StageExecutionResult[15:16] with 1 devices
StageExecutionResult[17:19] with 1 devices
StageExecutionResult[20:21] with 1 devices
StageExecutionResult[22:23] with 1 devices
StageExecutionResult[24:25] with 1 devices
StageExecutionResult[0:3] with 1 devices
StageExecutionResult[4:6] with 1 devices
StageExecutionResult[7:9] with 1 devices
StageExecutionResult[10:11] with 1 devices
StageExecutionResult[12:13] with 1 devices
StageExecutionResult[14:15] with 1 devices
StageExecutionResult[16:17] with 1 devices
StageExecutionResult[18:20] with 1 devices
StageExecutionResult[21:22] with 1 devices
StageExecutionResult[23:24] with 1 devices
StageExecutionResult[25:25] with 1 devices
StageExecutionResult[0:3] with 1 devices
StageExecutionResult[4:6] with 1 devices
StageExecutionResult[7:8] with 1 devices
StageExecutionResult[9:10] with 1 devices
StageExecutionResult[11:12] with 1 devices
StageExecutionResult[13:14] with 1 devices
StageExecutionResult[15:16] with 1 devices
StageExecutionResult[17:18] with 1 devices
StageExecutionResult[19:20] with 1 devices
StageExecutionResult[21:22] with 1 devices
StageExecutionResult[23:24] with 1 devices
StageExecutionResult[25:25] with 1 devices
StageExecutionResult[0:2] with 1 devices
StageExecutionResult[3:4] with 1 devices
StageExecutionResult[5:6] with 1 devices
StageExecutionResult[7:8] with 1 devices
StageExecutionResult[9:10] with 1 devices
StageExecutionResult[11:12] with 1 devices
StageExecutionResult[13:14] with 1 devices
StageExecutionResult[15:16] with 1 devices
StageExecutionResult[17:18] with 1 devices
StageExecutionResult[19:20] with 1 devices
StageExecutionResult[21:22] with 1 devices
StageExecutionResult[23:24] with 1 devices
StageExecutionResult[25:25] with 1 devices
[2024-09-12 20:42:45,453] [INFO] [worker.py:27:worker_main] Initializing torch.distributed...
[2024-09-12 20:42:45,454] [DEBUG] [engine.py:725:initialize_distributed] regenerate rank_map {'172.21.0.42:2220': [0], '172.21.0.42:2221': [1], '172.21.0.42:2222': [2], '172.21.0.42:2223': [3], '172.21.0.46:2220': [4], '172.21.0.46:2221': [5], '172.21.0.46:2222': [6], '172.21.0.46:2223': [7], '172.21.0.47:2220': [8], '172.21.0.47:2221': [9], '172.21.0.47:2222': [10], '172.21.0.47:2223': [11], '172.21.0.90:2220': [12]}
[2024-09-12 20:42:45,454] [INFO] [engine.py:739:initialize_distributed] init pg: rank 0, world_size: 13, rank_map: {'172.21.0.42:2220': [0], '172.21.0.42:2221': [1], '172.21.0.42:2222': [2], '172.21.0.42:2223': [3], '172.21.0.46:2220': [4], '172.21.0.46:2221': [5], '172.21.0.46:2222': [6], '172.21.0.46:2223': [7], '172.21.0.47:2220': [8], '172.21.0.47:2221': [9], '172.21.0.47:2222': [10], '172.21.0.47:2223': [11], '172.21.0.90:2220': [12]}
[2024-09-12 20:42:45,454] [INFO] [engine.py:750:initialize_distributed] Creating a TCP store on port: 33083
[2024-09-12 20:42:45,455] [INFO] [agent.py:232:forward_worker_port] Received worker port: 33083 from worker. Forwarding it to master...
[2024-09-12 20:42:45,456] [DEBUG] [agent.py:298:on_receive_response] Receiving: (<Response.FORWARD_RANK0_PORT: 5>, <RequestType.UNDEFINED: 0>)
[2024-09-12 20:42:45,457] [DEBUG] [agent.py:243:on_receive_worker_port] agent recv TCP Store port 33083 from master
[2024-09-12 20:42:46,324] [INFO] [comm.py:637:init_distributed] cdb=None
[2024-09-12 20:42:46,325] [INFO] [engine.py:775:initialize_distributed] [rank: 0] Distributed initialization is done.
[2024-09-12 20:42:46,325] [INFO] [worker.py:34:worker_main] Instantiating pipelines...
The objective value -7.4941E-09 of feasibility problem is less than zero_tolerance. This indicates that the nlp subproblem is feasible, although it is found infeasible in the previous step. Check the nlp solver output
[2024-09-12 20:42:46,637] [INFO] [instantiator.py:382:_distribute_batch] Batch distribution find no results. return None
The objective value -7.4941E-09 of feasibility problem is less than zero_tolerance. This indicates that the nlp subproblem is feasible, although it is found infeasible in the previous step. Check the nlp solver output
[2024-09-12 20:42:47,171] [INFO] [instantiator.py:382:_distribute_batch] Batch distribution find no results. return None
The objective value -7.4941E-09 of feasibility problem is less than zero_tolerance. This indicates that the nlp subproblem is feasible, although it is found infeasible in the previous step. Check the nlp solver output
The objective value -7.4941E-09 of feasibility problem is less than zero_tolerance. This indicates that the nlp subproblem is feasible, although it is found infeasible in the previous step. Check the nlp solver output
The objective value -7.4941E-09 of feasibility problem is less than zero_tolerance. This indicates that the nlp subproblem is feasible, although it is found infeasible in the previous step. Check the nlp solver output
[2024-09-12 20:42:48,376] [INFO] [instantiator.py:382:_distribute_batch] Batch distribution find no results. return None
The objective value -7.4941E-09 of feasibility problem is less than zero_tolerance. This indicates that the nlp subproblem is feasible, although it is found infeasible in the previous step. Check the nlp solver output
[2024-09-12 20:42:48,721] [INFO] [instantiator.py:382:_distribute_batch] Batch distribution find no results. return None
The objective value -7.4941E-09 of feasibility problem is less than zero_tolerance. This indicates that the nlp subproblem is feasible, although it is found infeasible in the previous step. Check the nlp solver output
[2024-09-12 20:42:49,025] [INFO] [instantiator.py:382:_distribute_batch] Batch distribution find no results. return None
The objective value -7.4941E-09 of feasibility problem is less than zero_tolerance. This indicates that the nlp subproblem is feasible, although it is found infeasible in the previous step. Check the nlp solver output
[2024-09-12 20:42:49,615] [INFO] [instantiator.py:382:_distribute_batch] Batch distribution find no results. return None
The objective value -7.4941E-09 of feasibility problem is less than zero_tolerance. This indicates that the nlp subproblem is feasible, although it is found infeasible in the previous step. Check the nlp solver output
The objective value 2.5059E-09 of feasibility problem is less than zero_tolerance. This indicates that the nlp subproblem is feasible, although it is found infeasible in the previous step. Check the nlp solver output
[2024-09-12 20:42:50,178] [INFO] [instantiator.py:382:_distribute_batch] Batch distribution find no results. return None
The objective value -7.4941E-09 of feasibility problem is less than zero_tolerance. This indicates that the nlp subproblem is feasible, although it is found infeasible in the previous step. Check the nlp solver output
[2024-09-12 20:42:50,529] [INFO] [instantiator.py:382:_distribute_batch] Batch distribution find no results. return None
The objective value -7.4941E-09 of feasibility problem is less than zero_tolerance. This indicates that the nlp subproblem is feasible, although it is found infeasible in the previous step. Check the nlp solver output
[2024-09-12 20:42:51,329] [INFO] [instantiator.py:382:_distribute_batch] Batch distribution find no results. return None
The objective value -7.4941E-09 of feasibility problem is less than zero_tolerance. This indicates that the nlp subproblem is feasible, although it is found infeasible in the previous step. Check the nlp solver output
[2024-09-12 20:42:55,307] [INFO] [instantiator.py:256:get_best_execution_plan] Best execution plan: 2 x <oobleck.PipelineTemplate.2nodes> pipelines (b: 11)
3 x <oobleck.PipelineTemplate.3nodes> pipelines (b: 14)
B: 64
dataloader: mbs 16
[2024-09-12 20:43:10,885] [INFO] [engine.py:159:_reconfiguration_listener_fn] ReconfigureEngine: start reconfigure listening
[2024-09-12 20:43:10,885] [INFO] [worker.py:36:worker_main] Begin training...
step: 0
step in try: 0
load_mbs: 0.0243074893951416, forward time: 7.257500410079956, backward time: 17.563080072402954
load_mbs cnt: 11, forward cnt: 11, backward cnt: 11
training step: 0 done
[2024-09-12 20:49:54,176] [INFO] [logging.py:96:log_dist] [Rank 0] time (ms) | step: 403292.16
/root/miniconda3/envs/oobleck/lib/python3.10/site-packages/torch/cuda/memory.py:416: FutureWarning: torch.cuda.memory_cached has been renamed to torch.cuda.memory_reserved
  warnings.warn(
/root/miniconda3/envs/oobleck/lib/python3.10/site-packages/torch/cuda/memory.py:424: FutureWarning: torch.cuda.max_memory_cached has been renamed to torch.cuda.max_memory_reserved
  warnings.warn(
[2024-09-12 20:49:54,179] [INFO] [logging.py:96:log_dist] [Rank 0]  | mem_allocated: 5.7512 GB | max_mem_allocated: 9.5473 GB | cache_allocated: 11.3281 GB | max_cache_allocated: 11.3281 GB
step: 1
step in try: 1
load_mbs: 0.01985788345336914, forward time: 6.8238348960876465, backward time: 17.666317462921143
load_mbs cnt: 11, forward cnt: 11, backward cnt: 11
training step: 1 done
[2024-09-12 20:56:26,081] [INFO] [logging.py:96:log_dist] [Rank 0] time (ms) | step: 391904.25
[2024-09-12 20:56:26,082] [INFO] [logging.py:96:log_dist] [Rank 0]  | mem_allocated: 5.7512 GB | max_mem_allocated: 12.3774 GB | cache_allocated: 14.3828 GB | max_cache_allocated: 14.3828 GB
step: 2
step in try: 2
load_mbs: 0.01985025405883789, forward time: 6.822565078735352, backward time: 17.5843505859375
load_mbs cnt: 11, forward cnt: 11, backward cnt: 11
training step: 2 done
